### ISSUE TITLE
Saving big array to file from parallel batch experiment doesn't work properly

### ISSUE THREAD

**Issue by curiosity29** [gama-platform/gama#3601]
**Describe the bug**
I tried to save some arrays to file using a parallel batch experiment like below but the result isn't as expected. Though it works fine with the non-parallel version and I haven't encountered any issue with a "small" array. My guess is because multiple threads were trying to save at the same time but I can't think of any workaround (Is there some kind of critical section in GAMA?)

**To Reproduce**
Using this model
```
model Experiment1dynamic

global
{
	int index_repeat <- 0;
	list list_4;
	list list_5;
	list list_6;
	int cycle_max <- 10000;
	reflex concat
	{
		list_4 <+ 4;
		list_5 <+ 5;		
		list_6 <+ 6;
	}
	string fileName <- "../includes/results/Capacity/" + "fish_variance_" +"test2"+ ".csv";
	reflex save_result when: cycle = cycle_max
	{
		save [index_repeat] + list_4 + list_5 + list_6
		to: fileName type: "csv" rewrite: false header: false;
	}
}
experiment "movement to variance" type: batch keep_seed: false until: cycle > cycle_max parallel: true
{
	parameter "repeat" var: index_repeat <- 1 among: [1, 2, 3];
}

```

**Expected behavior**
This should be the expected file (result from the non-parallel version):
![expected](https://user-images.githubusercontent.com/62107278/216901759-1d88d589-529a-4912-bbcd-95e57ca2bb24.png)

**Screenshots**
The actual file produced from the code:
![1](https://user-images.githubusercontent.com/62107278/216901557-84fba610-0174-4474-ae4d-899d94f49754.png)

**GAMA version: 1.9**

**Additional context**
Can I ask for another suggestion for #3575 as the suggestion of using an abort reflex for the simulation in doesn't seem to work (does not execute at the end of the simulation).

**Comment by AlexisDrogoul** [comment]
Hi,

Well noted for the bug. `parallel` is actually a completely experimental feature, and you should not use it everywhere. Best would be to use it only when it is critical (and not an issue), for example in the `ask` statement to ask agents to compute things in parallel (while their species still schedule them in sequence). 

Regarding the second point (#3575), please open a separate issue regarding `abort`.

**Comment by curiosity29** [comment]
It is probably not critical in most cases but definitely runs much faster as the normal version is not utilizing the computer resource. I tried to run with 4 instances of GAMA and it was using more memory than running 12 threads for 12 simulations. For the `abort`, I didn't open an issue because I was not sure if it just "works differently in this case" like the reflex in #3575 and the documentation for it seems to be missing so I don't know how it should work.

**Comment by AlexisDrogoul** [comment]
Identify the portions where it makes things run faster and put them inside a `ask ... parallel: true`.

For instance, instead of having : 

```
species fish parallel: true {
    reflex {
        do something;
    }

    .... do other things ... 
```

try 

```
global {

   reflex {
        ask fish parallel: true {
            do something;
        }
    }
}

species fish {

....
}
```

**Comment by curiosity29** [comment]
> try
> 
> ```
> global {
> 
>    reflex {
>         ask fish parallel: true {
>             do something;
>         }
>     }
> }
> 
> species fish {
> 
> ....
> }
> ```

I thought that specifying the species to be parallel mean that everything should be executed in parallel?
Also, using the above code, I get `ConcurrentModificationException: null` when using `at_distance` operator, not sure if it's the same bug as in #3576 or not.

**Comment by AlexisDrogoul** [comment]
Hi -- exactly. Not everything needs to be executed in parallel. And not everything can work in parallel too. Do not forget that using parallel in your models means you won't be able to reproduce/replicate your simulations. Depending on what you intend to do with them, it might be a big problem. Regarding the use of spatial operators, I will check, but have you implemented the fix provided in #3576 ?

**Comment by curiosity29** [comment]
> Hi -- exactly. Not everything needs to be executed in parallel. And not everything can work in parallel too. Do not forget that using parallel in your models means you won't be able to reproduce/replicate your simulations. Depending on what you intend to do with them, it might be a big problem. Regarding the use of spatial operators, I will check, but have you implemented the fix provided in #3576 ?

It should be the same as non-parallel if the process executed in parallel is independent of each other. Anyway, I feel like at least the reflex should be set to parallel when you set the species to be parallel because it's the same as asking the species to do something in parallel.
Do you mean the fix using `gama.pref_synchronize_quadtree <- true;` and moving the reflex of fish_killer to global like this? It doesn't seem to fix the problem. And I'm not sure what you mean by letting the simulation do the at_distance operator since that operator needs the caller to be the fish.
```
model test

global
{
	int max_fish <- 1000;
	init
	{
		seed<-10.0;
		gama.pref_synchronize_quadtree <- true;
//		create fish_killer;
		create fish number: max_fish;
	}

	reflex kill
		{
			loop f over: fish 
			{
				if f.dead = false and flip (0.5)
				{
					f.dead <- true;
				}
			}
		}
	int stop_at <- 100;
	reflex stop when: cycle > stop_at
	{
		do pause;
	}
	reflex do_die
	{
		ask fish
		{
			if dead = true{ do die;}
		}
	}
	reflex write_friend
	{
		write "cycle:" + cycle;
		
		ask one_of(fish)
		{
			write "number of fish at distance: " + (length(fish at_distance 200) + 1);
			write "number of total fish: " + length(fish);
		}
	}
}
species fish parallel: true
{
	bool dead <- false;
	aspect base 
	{
		draw circle(0.5) color: #red;
	}
	int friends update: length(fish at_distance 200.0) +1;

	reflex make_friends 
	{
		if max_fish > friends
		{
			create fish;
		}
	}
}

experiment main
{
	output
	{
	display map
		{
			species fish aspect: base;
		}
	}
}
```

**Comment by AlexisDrogoul** [comment]
Hi,

Coming back to this, what I mean is that you cannot expect a sequential simulation to behave like a parallel simulation because there are factors (like the scheduling of threads) that GAMA cannot control and that might have a consequent effect on the agents. The way you can control this, however, is to synchronise yourself some parts of the code by running them in the simulation (`global`) rather than in each agent. Especially when it concerns the creation / deletion of agents. So for instance, instead of having a reflex like this in the fishes: 

```
species fish parallel: true
{
	bool dead <- false;
	aspect base 
	{
		draw circle(0.5) color: #red;
	}
	int friends update: length(fish at_distance 200.0) +1;

	reflex make_friends 
	{
		if max_fish > friends
		{
			create fish;
		}
	}
}
```

where you are likely to run into trouble at one point, I would tend, in a parallel setup, to privilegiate something like:

```
global {
     ...
     reflex build_friends {
          map<fish,int> friends <- fish as_map (each, length(each neighbors_at 200)); 
          ask fish {
                   if (max_fish > friends[self]) { create fish;}
          }
     }
```

I did not test it (just wrote the text here) but you get the idea: in a parallel setup, when you have critical operations like creating or counting agents, you need to resynchronise things yourself (like you would do in a critical section).
